### PR TITLE
Partial current tallies as cell_to_cell tallies

### PR DESCRIPTION
**Pull request by GiudGiud**
Hello, 

This PR is meant to address issue #2  as well as help me keep openmc up to date during my PhD. It's my 2nd PR so I probably missed a lot of things I should do/have done, but let's keep learning.

This PR adds partial current tallies as a combination of 2 filters:
- cell_from filter, a filter on the cell from which the particle comes from when it's crossing a surface
- cell_to filter, a filter on the cell which the particle enters when it's crossing a surface

In addition, it creates 2 scores:
- SCORE_CELL_TO_CELL: just counts particles, like an event score
- SCORE_CELL_TO_CELL_NORMAL_PROJECTION: scores the cosine of the angle between the particle direction and the normal of the surface

It also includes its **own list of tallies**, to avoid looking at non-relevant tallies when crossing surfaces: active_cell_to_cell_tallies

The score_partial_current subroutine is called when crossing a surface (subroutine cross_surface), it loops through the tallies in the list, all the bins etc, and scores according to what score one wants. 

Why not a surface filter?
-Surface filters make it very complicated to tally currents in a rings and sectors discretization, that I am using in my MOC solver later on. Even with a surface filter, one would still need a cell_to or cell_from filters to do partial currents between sectors. A surface is typically shared between many cells. 
-Also, when going from MC to MOC, it is much better to work with cells than with surfaces, since MOC tracks have no idea which surface they're crossing when they're crossing one. 
-Surface filters are nice for other things, it'd be ok to have the two.

**I wish to know** if you are interested in having this capability of tallying partial currents done this way or similarly. If you are, I can complete this PR to include tests, rename everything nicely, etc.

What this PR doesn't have / can't do but that I could add in the future (depending on my needs / the needs of the community):
- handling of boundary conditions, not sure what I would need to do yet
- tallying leakage rates (can still be done with combinations of cell_from and cell_to)
- tests

What this PR has:
- a very general implementation of cell_to_filter. Mainly because it shares a lot of code with the regular cellfilter. 
- cell_from can only have one member right now. I don't need the capability for more but I can add it if you want it for the PR or later. If you want it, to be the most general I'm going to need to add a particle % last_coord to save the last coordinates at all levels of the particle
- a very general implementation of score_partial_current. Mainly because it shares a lot of code with other scoring subroutines. 
- it is a surface integrated tally.
- input/output stuff

### CODE REVIEW

**Comment by GiudGiud** [comment]
oups, sorry for language in 91d2a15

File: src/tally.F90 [hunk 1]
```diff
@@ -3138,6 +3149,128 @@ subroutine score_collision_tally(p)
...
+          i_filt = t % filter(j)
+          i_bin = filter_matches(i_filt) % i_bin
+          filter_index = filter_index + (filter_matches(i_filt) % bins % &
+               data(i_bin) - 1) * t % stride(j)
+          filter_weight = filter_weight * filter_matches(i_filt) % weights % &
+               data(i_bin)
+        end do
+
+        call score_general(p, t, 0, filter_index, &
+             0, 0d0, flux * filter_weight)
```
Comment (smharper) [inline src/tally.F90]: It doesn't make sense for this function to call `score_general`.  There would be no reason to have say a `scatter` score that is only called when a particle crosses a surface.  Just put the relevant code from `score_general` directly in this function.

File: src/geometry.F90 [hunk 2]
```diff
@@ -610,13 +611,33 @@ subroutine cross_surface(p, last_cell)
...
       call find_cell(p, found, surf%neighbor_pos)
+
+      ! Save cosine of angle between surface normal and particle direction
+      p % normal_proj = dot_product(p % coord(1) % uvw, &
+                  & surf % normal(p % coord(1) % xyz)) / &
+                  & sqrt(dot_product(surf % normal(p % coord(1) % xyz)&
+                  &, surf % normal(p % coord(1) % xyz)))
+
+      ! Score cell to cell partial currents
+      call score_partial_current(p)
```
Comment (GiudGiud) [inline src/geometry.F90]: Good idea. I put it there to be with score_partial_current, but it s less code in less locations to have it in tracking

File: src/tally_filter.F90 [hunk 3]
```diff
@@ -72,6 +72,32 @@ module tally_filter
...
+    procedure :: get_next_bin => get_next_bin_cell_from
+    procedure :: to_statepoint => to_statepoint_cell_from
+    procedure :: text_label => text_label_cell_from
+    procedure :: initialize => initialize_cell_from
+  end type CellFromFilter
+
+!===============================================================================
+! CELLTOFILTER specifies which geometric cells particles enter.
+!===============================================================================
+  type, extends(TallyFilter) :: CellToFilter
```
Comment (GiudGiud) [inline src/tally_filter.F90]: ok did that. Did the same in CellFromFilter

File: src/tally_filter.F90 [hunk 4]
```diff
@@ -718,6 +744,168 @@ function text_label_cell(this, bin) result(label)
...
+  end subroutine get_next_bin_cell_from
+
+  subroutine to_statepoint_cell_from(this, filter_group)
+    class(CellFromFilter), intent(in) :: this
+    integer(HID_T),        intent(in) :: filter_group
+
+    integer :: i
+    integer, allocatable :: cell_ids(:)
+
+    call write_dataset(filter_group, "type", "cell")
```
Comment (GiudGiud) [inline src/tally_filter.F90]: oups

**Comment by GiudGiud** [comment]
Thanks for the review! Great ideas on making it more general and use less code. I implemented each of them. 
The cell from filter is now completely general, it can handle more than one bin, and at multiple levels.
We could consider copying less data in last_coord, for performance

**Comment by smharper** [comment]
I think the basics of the code here are fine, and I'm willing to merge it when all the little details are figured out.  Go ahead and update the tests and docs.

@paulromano, the part in here that worries me most is adding the new set of `LocalCoord`s to the `Particle`.  Do you know off-hand if our performance is sensitive to the size of `Particle`?  (Even if it isn't now, I wonder if it might become important for DD.)  If so, we could either reduce the burden by only storing cell ids or maybe passing the relevant info directly to `score_partial_current` from tracking.F90

File: src/particle_header.F90 [hunk 5]
```diff
@@ -217,6 +226,48 @@ subroutine initialize_from_source(this, src, run_CE, energy_bin_avg)
...
+
+    class(Particle), intent(inout) :: this
+    integer :: i
+
+    this % last_n_coord = this % n_coord
+
+    ! copy all information at all coordinate levels
+    do i = 1, this % n_coord
+      call this % last_coord(i) % copy_coord(this % coord(i))
+    end do
```
Comment (paulromano) [inline src/particle_header.F90]: Might just be simpler to just do `this % last_coord = this % coord`. Although it would result in all the xyz and uvw data getting copied too, the compiler might be able to better optimize a straight memory copy.

File: src/tally.F90 [hunk 6]
```diff
@@ -3138,6 +3138,146 @@ subroutine score_collision_tally(p)
...
 
 !===============================================================================
+! score_partial_current tallies partial currents from cell_to to cell_from
+!===============================================================================
+
+    subroutine  score_partial_current(p)
+
+    type(Particle), intent(in)    :: p
+
+  !end if
```
Comment (paulromano) [inline src/tally.F90]: remove this

File: src/tally.F90 [hunk 7]
```diff
@@ -3138,6 +3138,146 @@ subroutine score_collision_tally(p)
...
+    integer :: score_bin            ! scoring bin, e.g. SCORE_FLUX
+    integer :: score_index          ! scoring bin index
+    integer :: j                    ! loop index for scoring bins
+    integer :: filter_index         ! single index for single bin
+    integer :: matching_bin         ! next valid filter bin
+    real(8) :: flux                 ! collision estimate of flux
+    real(8) :: filter_weight        ! combined weight of all filters
+    real(8) :: score                ! analog tally score
+    logical :: finished             ! found all valid bin combinations
+    type(TallyObject), pointer :: t
```
Comment (paulromano) [inline src/tally.F90]: Get rid of this and just use an associate block, i.e.
```Fortran
associate (t => tallies(i_tally))
   ...
end associate
```

File: src/tracking.F90 [hunk 8]
```diff
@@ -158,9 +167,12 @@ subroutine transport(p)
         else
           ! Particle crosses surface
           p % surface = surface_crossed
+
           call cross_surface(p, last_cell)
           p % event = EVENT_SURFACE
         end if
+        ! Score cell to cell partial currents
+        if(active_cell_to_cell_tallies%size()>0) call score_partial_current(p)
```
Comment (paulromano) [inline src/tracking.F90]: Style fixes:
`if (active_cell_to_cell_tallies % size() > 0) call score_partial_current(p)`

File: src/particle_header.F90 (hunk 5 above)
Comment (GiudGiud) [inline src/particle_header.F90]: Will that actually copy all the data or just a pointer?

Comment (paulromano) [inline src/particle_header.F90]: It will actually copy all the data since the variables do not have the `pointer` attribute.

**Comment by GiudGiud** [comment]
Thanks for the review. I'm glad I can contribute. I like the new set of names. I'll get them in with the tests and documentation. 

I didn't try a surface filter. I'll post the result once I've tried.
What I have tried is to use a surface current tally with the cell_to and cell_from filters. This doesn't work, only energy mesh and surface filters have been implemented in surface current tallies.

File: src/particle_header.F90 (hunk 5 above)
Comment (GiudGiud) [inline src/particle_header.F90]: Ok thanks. I ll try that.

**Comment by GiudGiud** [comment]
Surface filters currently do not work with the new surface tally. But it's pretty close to being functional.
The problem is that get_next_surface compares p%surface and this (the filter) % surfaces(i) and:
- this % surfaces(i) is the surface id, so always positive
- p%surface is positive when the particle comes from one side of the surface, and negative on the other side

The result is that only particles crossing on the "+ side" get scored. So the behavior is not exactly as expected. Indeed, it becomes impossible to tally particles entering and exiting a cylinder for example. 
I see multiple options for solving this:
- allowing negative integer bins in surface filters and adding absolute values when openmc checks that the surface exists
- modifying get_next_bin_surface to compare absolute values, but this is going to break cmfd/mesh tallies I bet

What do you prefer?

Otherwise I've created a new test file for the tally and added the tallies and scores in the documentation. Should I also add an example?

**Comment by GiudGiud** [comment]
I've gone ahead and implemented the second option to allow surface tallies to be used. 
This means that surface tallies naturally give **total currents**, not partial or net currents.
 
In my opinion, this is better because it allows for **scores** to determine if a tally should obtain a
- partial current
- net current
- total current
In consequence, I have renamed my score "total current" instead of "partial current". 
The partial and net current scores are not in the PR. Partial currents can still be obtained without a new score by using cell_from and/or cell(to) filters. 

However, be advised that in the case of nested cells, cubes inside cubes for example, in the outer cube an event is tallied every time the surfaces defining the inner cube is crossed. As opposed to only when the particle goes from the outer to the inner cube. 

I don't know why Travis fails to build. Anyone does?

File: docs/source/io_formats/tallies.rst [hunk 9]
```diff
@@ -154,14 +154,30 @@ For each filter type, the following table describes what the ``bins`` attribute
 should be set to:
 
 :cell:
-  A list of unique IDs for cells in which the tally should be accumulated.
+  A list of cells or unique IDs for cells in which the tally should be
```
Comment (paulromano) [inline docs/source/io_formats/tallies.rst]: Since this is describing XML, the original wording was correct, i.e., it should be a list of unique IDs for cells.

File: docs/source/io_formats/tallies.rst [hunk 10]
```diff
@@ -154,14 +154,30 @@ For each filter type, the following table describes what the ``bins`` attribute
...
+
+:surface:
+  This filter allows the tally to be scored when crossing a surface. A list of 
+  surface IDs should be given. It does not specify in which direction the 
+  surface is crossed, and a cellfrom or a cell filter may be used to tally 
+  partial currents.
+
+:cellfrom:
+  This filter allows the tally to be scored when crossing a surface and the 
+  particle came from a specified cell. A list of cell or cell IDs should be
```
Comment (paulromano) [inline docs/source/io_formats/tallies.rst]: a list of cell IDs

File: docs/source/io_formats/tallies.rst [hunk 11]
```diff
@@ -154,14 +154,30 @@ For each filter type, the following table describes what the ``bins`` attribute
...
+  particle came from a specified cell. A list of cell or cell IDs should be 
+  given.
+  To tally a partial current from a cell to another, this filter should be 
+  used in combination with a cell filter, to define the other cell.
+  This filter should not be used in combination with a meshfilter.
 
 :cellborn:
   This filter allows the tally to be scored to only when particles were
-  originally born in a specified cell. A list of cell IDs should be given.
+  originally born in a specified cell. A list of cell or cell IDs should be
```
Comment (paulromano) [inline docs/source/io_formats/tallies.rst]: a list of cell IDs

File: docs/source/io_formats/tallies.rst [hunk 12]
```diff
@@ -154,14 +154,30 @@ For each filter type, the following table describes what the ``bins`` attribute
...
+  surface IDs should be given. It does not specify in which direction the 
+  surface is crossed, and a cellfrom or a cell filter may be used to tally 
+  partial currents.
+
+:cellfrom:
+  This filter allows the tally to be scored when crossing a surface and the 
+  particle came from a specified cell. A list of cell or cell IDs should be 
+  given.
+  To tally a partial current from a cell to another, this filter should be 
+  used in combination with a cell filter, to define the other cell.
```
Comment (paulromano) [inline docs/source/io_formats/tallies.rst]: Maybe reword to "To tally a partial current from one cell to another, this filter should be used in combination with a cell filter that defines the other cell."

**Comment by smharper** [comment]
I can't think of a use-case for a total current, so let's have the `SurfaceFilter` default to net current.  We can implement that by setting the `weight = sign(ONE, this % surfaces(i))`.  If a user wants partial current instead of net, they can just go to `CellFilter` and `CellFromFilter`

**Comment by smharper** [comment]
Correction: `weight = sign(ONE, p % surface)`

**Comment by smharper** [comment]
I think it should be *filters* not *scores* that differentiate between net and partial current.  The implementation you have now with `CellFromFilter` already fits that abstraction.  Consider also if we wanted to implement a partial current tally for surfaces.  There would be two different bins in the results arrays per surface; one for the positive current and one for the negative current.  It fits our existing abstractions better for a filter to find which of those bins is appropriate rather than a score.

So I suggest getting rid of the `case('total-current')` in input_xml.F90 and just reusing `case('current')`.  That `case` block would then look at the filters and decide the `t % type` based on whether a mesh filter, surface filter, or cell-from filter was given.  You'd need to add some code to call `fatal_error` if more than one of those filter types is given.

The constants `TALLY_SURFACE_CURRENT` and `TALLY_SURFACE` are a little confusing.  Let's change `TALLY_SURFACE_CURRENT` to `TALLY_MESH_CURRENT`.

Also, I'd prefer that you use a `PyAPITestHarness` for the new test.  That way, we are checking the Python code in addition to the Fortran code.

**Comment by smharper** [comment]
Oh, and to answer your question about an example:  I don't think an example is necessary.  The documentation and the test should be sufficient.

**Comment by paulromano** [comment]
I second @smharper's comments. In MCNP, they have tallies for surface net currents, and if you want partial currents, they suggest using cosine bins which would be the equivalent of our PolarFilter. Of course, that only works for planar surfaces. We should eventually extend our PolarFilter to be able to specify a reference direction since right now it only works relative to the positive z-axis, so you could only get partial currents for z-planes.

**Comment by paulromano** [comment]
I worry slightly about the size of the `Particle` type. When I compared performance of your branch versus develop on a reactor problem, I see about a 4% performance decrease. To get that performance back, one thing we can do is just store the cells instead of the entire coordinates. My suggestion for how to do this is the following:
```diff
diff --git a/src/geometry.F90 b/src/geometry.F90
index f80bd7e..c451b61 100644
--- a/src/geometry.F90
+++ b/src/geometry.F90
@@ -384,9 +384,8 @@ contains
 ! the geometry, is reflected, or crosses into a new lattice or cell
 !===============================================================================
 
-  subroutine cross_surface(p, last_cell)
+  subroutine cross_surface(p)
     type(Particle), intent(inout) :: p
-    integer,        intent(in)    :: last_cell  ! last cell particle was in
 
     real(8) :: u          ! x-component of direction
     real(8) :: v          ! y-component of direction
@@ -469,7 +468,7 @@ contains
       p%coord(1)%uvw(:) = [u, v, w] / norm
 
       ! Reassign particle's cell and surface
-      p % coord(1) % cell = last_cell
+      p % coord(1) % cell = p % last_cell(p % last_n_coord)
       p % surface = -p % surface
 
       ! If a reflective surface is coincident with a lattice or universe
diff --git a/src/particle_header.F90 b/src/particle_header.F90
index 7fcdf56..76876f0 100644
--- a/src/particle_header.F90
+++ b/src/particle_header.F90
@@ -49,8 +49,8 @@ module particle_header
     type(LocalCoord) :: coord(MAX_COORD) ! coordinates for all levels
 
     ! Particle coordinates before crossing a surface
-    integer          :: last_n_coord          ! number of current coordinates
-    type(LocalCoord) :: last_coord(MAX_COORD) ! coordinates for all levels
+    integer :: last_n_coord         ! number of current coordinates
+    integer :: last_cell(MAX_COORD) ! coordinates for all levels
 
     ! Energy Data
     real(8)    :: E      ! post-collision energy
@@ -110,7 +110,6 @@ module particle_header
     procedure :: clear => clear_particle
     procedure :: initialize_from_source
     procedure :: create_secondary
-    procedure :: set_last_coord
   end type Particle
 
 contains
@@ -152,7 +151,6 @@ contains
     ! Set up base level coordinates
     this % coord(1) % universe = root_universe
     this % n_coord = 1
-    this % last_coord(1) % universe = root_universe
     this % last_n_coord = 1
 
   end subroutine initialize_particle
@@ -225,26 +223,6 @@ contains
   end subroutine initialize_from_source
 
 !===============================================================================
-! SET_LAST_COORD copies all coordinate levels from coord to last_coord. This is
-! meant to have information about the last cell in the cell_from filter
-!===============================================================================
-
-  subroutine set_last_coord(this)
-
-    class(Particle), intent(inout) :: this
-    integer :: i
-
-    this % last_n_coord = this % n_coord
-    this % last_coord = this % coord
-
-    ! reset the rest of former last_coord
-    do i = this % n_coord + 1, MAX_COORD
-      call this % last_coord(i) % reset()
-    end do
-
-  end subroutine set_last_coord
-
-!===============================================================================
 ! CREATE_SECONDARY stores the current phase space attributes of the particle in
 ! the secondary bank and increments the number of sites in the secondary bank.
 !===============================================================================
diff --git a/src/tally_filter.F90 b/src/tally_filter.F90
index 6dc3310..a19a6fe 100644
--- a/src/tally_filter.F90
+++ b/src/tally_filter.F90
@@ -746,7 +746,7 @@ contains
       start = 1
     else
       do i = 1, p % last_n_coord
-        if (p % last_coord(i) % cell == this % cells(current_bin)) then
+        if (p % last_cell(i) == this % cells(current_bin)) then
           start = i + 1
           exit
         end if
@@ -757,8 +757,8 @@ contains
     next_bin = NO_BIN_FOUND
     weight = ERROR_REAL
     do i = start, p % last_n_coord
-      if (this % map % has_key(p % last_coord(i) % cell)) then
-        next_bin = this % map % get_key(p % last_coord(i) % cell)
+      if (this % map % has_key(p % last_cell(i))) then
+        next_bin = this % map % get_key(p % last_cell(i))
         weight = ONE
         exit
       end if
diff --git a/src/tracking.F90 b/src/tracking.F90
index 8e67843..546ae38 100644
--- a/src/tracking.F90
+++ b/src/tracking.F90
@@ -36,7 +36,6 @@ contains
     integer :: next_level             ! next coordinate level to check
     integer :: surface_crossed        ! surface which particle is on
     integer :: lattice_translation(3) ! in-lattice translation vector
-    integer :: last_cell              ! most recent cell particle was in
     integer :: n_event                ! number of collisions/crossings
     real(8) :: d_boundary             ! distance to nearest boundary
     real(8) :: d_collision            ! sampled distance to collision
@@ -150,8 +149,10 @@ contains
         if (next_level > 0) p % n_coord = next_level
 
         ! Saving previous cell data
-        last_cell = p % coord(p % n_coord) % cell
-        call p % set_last_coord()
+        do j = 1, p % n_coord
+          p % last_cell(j) = p % coord(j) % cell
+        end do
+        p % last_n_coord = p % n_coord
 
         ! Update last_ data. This is needed to use the same filters in
         ! surface tallies as the ones implemented for regular tallies
@@ -168,7 +169,7 @@ contains
           ! Particle crosses surface
           p % surface = surface_crossed
 
-          call cross_surface(p, last_cell)
+          call cross_surface(p)
           p % event = EVENT_SURFACE
         end if
         ! Score cell to cell partial currents
```

File: docs/source/io_formats/tallies.rst (hunk 9 above)
Comment (GiudGiud) [inline docs/source/io_formats/tallies.rst]: My bad. Still not used to not seeing the Python API as the only way to input parameters

**Comment by GiudGiud** [comment]
Thanks for the review! I see how filters can help us with getting both partial currents more easily. 
We're back to keeping only the last_cell for each particle surface crossing. I hope that gets us all the performance back!

I'm not done with the tests yet. It may fail my new test for now as surface tallies have a different behavior.

File: src/tally_filter.F90 [hunk 13]
```diff
@@ -887,9 +961,9 @@ subroutine get_next_bin_surface(this, p, estimator, current_bin, next_bin, &
     weight = ERROR_REAL
     if (current_bin == NO_BIN_FOUND) then
       do i = 1, this % n_bins
-        if (p % surface == this % surfaces(i)) then
+        if (abs(p % surface) == this % surfaces(i)) then
           next_bin = i
-          weight = ONE
+          weight = sign(1, p % surface)
```
Comment (paulromano) [inline src/tally_filter.F90]: `weight = sign(ONE, p % surface)`

File: docs/source/usersguide/tallies.rst [hunk 14]
```diff
@@ -261,12 +261,19 @@ The following tables show all valid scores:
...
+    |                      |a mesh. It may not be used in conjunction with any |
+    |                      |other score. Only energy and mesh filters may be   |
+    |                      |used.                                              |
+    |                      |Used in combination with a surface filter:         |
+    |                      |Net currents on any surface previously defined in  |
+    |                      |the geometry. It may be used along with any other  |
+    |                      |filter, except mesh filters.                       |
+    |                      |Surfaces can alternatively be defined with cell    |
+    |                      |from and cell filters and partial currents can be  |
+    |                      |obtained by using cell_from and cell filters.      |
```
Comment (paulromano) [inline docs/source/usersguide/tallies.rst]: Maybe reword to "Surfaces can alternatively be defined with cellfrom and cell filters, thereby resulting in partial currents."

File: src/tally_filter.F90 (hunk 13 above)
Comment (smharper) [inline src/tally_filter.F90]: It looks like both arguments need to be of the same type ([see here](https://gcc.gnu.org/onlinedocs/gfortran/SIGN.html)) so in order to use `ONE` he'd need to cast `p % surface`.  Since there is a type cast involved no matter what, I think it's fine to stick with these integer arguments.

Comment (paulromano) [inline src/tally_filter.F90]: I stand corrected. Thanks @smharper!

**Comment by smharper** [comment]
@GiudGiud, if you're curious why check source gave you the wrong line number, it's because Travis first merges your code into develop before running the tests.  If you have trouble with this in the future, remember that you can always run the check_source script on your machine, and it should give the correct line numbers.

**Comment by GiudGiud** [comment]
Tests are passing on my machine

**Comment by paulromano** [comment]
@GiudGiud That's right -- what you should do is merge the develop branch into your branch and then re-generate results for test_surface_tally:
```bash
git merge develop
cd tests/test_surface_tally
python test_surface_tally.py --update
```

File: openmc/filter.py [hunk 15]
```diff
@@ -18,7 +18,8 @@
 
 _FILTER_TYPES = ['universe', 'material', 'cell', 'cellborn', 'surface',
                  'mesh', 'energy', 'energyout', 'mu', 'polar', 'azimuthal',
-                 'distribcell', 'delayedgroup', 'energyfunction']
+                 'distribcell', 'delayedgroup', 'energyfunction', 'cellfrom', 
+                 'cellto']
```
Comment (paulromano) [inline openmc/filter.py]: Get rid of this 'cell'to'

File: src/input_xml.F90 [hunk 16]
```diff
@@ -3004,7 +3004,7 @@ subroutine read_tallies_xml()
         end if
         n_words = node_word_count(node_filt, "bins")
       case ("mesh", "universe", "material", "cell", "distribcell", &
-            "cellborn", "surface", "delayedgroup")
+            "cellborn", "cellto", "cellfrom", "surface", "delayedgroup")
```
Comment (paulromano) [inline src/input_xml.F90]: Remove "cellto"

File: tests/test_surface_tally/test_surface_tally.py [hunk 17]
```diff
@@ -0,0 +1,164 @@
+#!/usr/bin/env python
+
+import os
+import sys
+sys.path.insert(0, os.pardir)
+from testing_harness import PyAPITestHarness
+import numpy as np
+import openmc
+import pandas as pd
+
```
Comment (paulromano) [inline tests/test_surface_tally/test_surface_tally.py]: PEP8: Need two blank lines before a class definition

File: tests/test_surface_tally/test_surface_tally.py [hunk 18]
```diff
@@ -0,0 +1,164 @@
...
+        borated_water.add_s_alpha_beta('c_H_in_H2O')
+
+        # Instantiate a Materials collection and export to XML
+        materials_file = openmc.Materials([uo2, borated_water])
+
+        materials_file.export_to_xml()
+
+        # Instantiate ZCylinder surfaces
+        fuel_or = openmc.ZCylinder(surface_id=1, x0=0, y0=0, R=0.4, name='Fuel OR')
+        left    = openmc.XPlane(surface_id=2, x0=-0.62992, name='left')
```
Comment (paulromano) [inline tests/test_surface_tally/test_surface_tally.py]: PEP8: don't use multiple spaces before operator, i.e. `left = openmc.XPlane(...)`

File: tests/test_surface_tally/test_surface_tally.py [hunk 19]
```diff
@@ -0,0 +1,164 @@
...
+        tallies_file = openmc.Tallies()
+
+        # Cell to cell tallies
+        # These filters are same for all tallies
+        two_groups = np.array([0., 4, 20.]) * 1e6
+        energy_filter     = openmc.EnergyFilter(two_groups)
+        polar_filter      = openmc.PolarFilter([0, np.pi / 4, np.pi])
+        azimuthal_filter  = openmc.AzimuthalFilter([0, np.pi / 4, np.pi])
+
+
```
Comment (paulromano) [inline tests/test_surface_tally/test_surface_tally.py]: Remove one blank line here

File: tests/test_surface_tally/test_surface_tally.py [hunk 20]
```diff
@@ -0,0 +1,164 @@
...
+        cell_to_cell_tallies = []
+        tally_index = 0
```
Comment (paulromano) [inline tests/test_surface_tally/test_surface_tally.py]: I wouldn't even bother assigning indices manually; just let the Python API assign them.

File: tests/test_surface_tally/test_surface_tally.py [hunk 21]
```diff
@@ -0,0 +1,164 @@
...
+        # Instantiate Cells
+        fuel = openmc.Cell(name='fuel')
+        water = openmc.Cell(name='water')
+
+        # Use surface half-spaces to define regions
+        fuel.region = -fuel_or
+        water.region = +fuel_or & -right & +bottom & -top
+
+        # Register Materials with Cells
+        fuel.fill  = uo2
```
Comment (paulromano) [inline tests/test_surface_tally/test_surface_tally.py]: Don't use multiple spaces before operator

File: tests/test_surface_tally/test_surface_tally.py [hunk 22]
```diff
@@ -0,0 +1,164 @@
...
+        settings_file.export_to_xml()
+
+        # Tallies file
+        tallies_file = openmc.Tallies()
+
+        # Cell to cell tallies
+        # These filters are same for all tallies
+        two_groups = np.array([0., 4, 20.]) * 1e6
+        energy_filter     = openmc.EnergyFilter(two_groups)
+        polar_filter      = openmc.PolarFilter([0, np.pi / 4, np.pi])
```
Comment (paulromano) [inline tests/test_surface_tally/test_surface_tally.py]: Don't use multiple spaces before operator

File: tests/test_surface_tally/test_surface_tally.py [hunk 23]
```diff
@@ -0,0 +1,164 @@
...
+                    cell_to_cell_tallies[tally_index].filters   = [cell_from_filter, surface_filter, energy_filter, polar_filter, azimuthal_filter]
+                    cell_to_cell_tallies[tally_index].scores    = ['current']
+                    tallies_file.append(cell_to_cell_tallies[tally_index])
+                    tally_index += 1
+
+        # Surface filter on inner surface, for net current
+        surface_filter = openmc.SurfaceFilter([1])
+        surf_tally1 = openmc.Tally(tally_id=tally_index, name='net_cylinder')
+        surf_tally1.filters   = [surface_filter, energy_filter, polar_filter, azimuthal_filter]
+        surf_tally1.scores    = ['current']
```
Comment (paulromano) [inline tests/test_surface_tally/test_surface_tally.py]: Don't use multiple spaces before operator

File: tests/test_surface_tally/test_surface_tally.py [hunk 24]
```diff
@@ -0,0 +1,164 @@
...
+        energy_filter     = openmc.EnergyFilter(two_groups)
+        polar_filter      = openmc.PolarFilter([0, np.pi / 4, np.pi])
+        azimuthal_filter  = openmc.AzimuthalFilter([0, np.pi / 4, np.pi])
+
+
+        cell_to_cell_tallies = []
+        tally_index = 0 
+
+        for cell1 in pin_cell.get_all_cells():
+            for cell2 in pin_cell.get_all_cells():
```
Comment (paulromano) [inline tests/test_surface_tally/test_surface_tally.py]: This double-loop seems overly complicated since there are only two cells in the `pin_cell` universe. You might as well just manually add the filters/tallies needed.

File: tests/test_surface_tally/test_surface_tally.py [hunk 25]
```diff
@@ -0,0 +1,164 @@
...
+        sp = openmc.StatePoint(self._sp_name)
+
+        # Extract the tally data as a Pandas DataFrame.
+        df = pd.DataFrame()
+        for t in sp.tallies.values():
+            df = df.append(t.get_pandas_dataframe(), ignore_index=True)
+
+        # Extract the relevant data as a CSV string.
+        cols = ('d_material', 'd_nuclide', 'd_variable', 'score', 'mean',
+                'std. dev.')
```
Comment (paulromano) [inline tests/test_surface_tally/test_surface_tally.py]: I would just make this `cols = ('mean', 'std. dev.')`. The first three variables don't apply to this test and 'score' is always 'current', so not very useful.

File: tests/test_surface_tally/test_surface_tally.py [hunk 26]
```diff
@@ -0,0 +1,164 @@
...
+
+import os
+import sys
+sys.path.insert(0, os.pardir)
+from testing_harness import PyAPITestHarness
+import numpy as np
+import openmc
+import pandas as pd
+
+class CreateSurfaceTallyTestHarness(PyAPITestHarness):
```
Comment (paulromano) [inline tests/test_surface_tally/test_surface_tally.py]: Maybe just `SurfaceTallyTestHarness`

File: tests/test_surface_tally/test_surface_tally.py (hunk 23 above)
Comment (GiudGiud) [inline tests/test_surface_tally/test_surface_tally.py]: My bad, forgot to reformat the test to the standard.

**Comment by GiudGiud** [comment]
Having trouble with the tests. 
The develop branch doesn't get me the right results anymore (since C API).
Investigating!

**Comment by paulromano** [comment]
Thanks for fixing this up @GiudGiud; I'm satisfied at this point. @smharper do you have any remaining requests before we merge?

**Comment by smharper** [comment]
No further requests.  Thanks for the addition, @GiudGiud!
